p2p/gossip_tracer.go

```diff
@@ -0,0 +1,218 @@
+		psRPCSubSent: promauto.NewCounter(prometheus.CounterOpts{
+			Name: "p2p_pubsub_rpc_sent_sub_total",
+			Help: "The number of subscription messages sent via rpc",
+		}),
+		psRPCPubSent: promauto.NewCounterVec(prometheus.CounterOpts{
+			Name: "p2p_pubsub_rpc_sent_pub_total",
+			Help: "The number of publish messages sent via rpc",
+		}, []string{"topic"}),
+	}
+}
+
+func (g *gossipTracer) AddPeer(p peer.ID, proto protocol.ID) {
+	g.p2PPeerCount.WithLabelValues(agentNameFromPeerID(p, g.host.Peerstore())).Inc()
+}
+
+func (g *gossipTracer) RemovePeer(p peer.ID) {
+	g.p2PPeerCount.WithLabelValues(agentNameFromPeerID(p, g.host.Peerstore())).Dec()
+}
+
+func (g *gossipTracer) Join(topic string) {
+	g.psTopicsActive.WithLabelValues(topic).Set(1)
+}
+
+func (g *gossipTracer) Leave(topic string) {
+	g.psTopicsActive.WithLabelValues(topic).Set(0)
+}
+
+func (g *gossipTracer) Graft(p peer.ID, topic string) {
+	g.psTopicsGraft.WithLabelValues(topic).Inc()
+}
+
+func (g *gossipTracer) Prune(p peer.ID, topic string) {
+	g.psTopicsPrune.WithLabelValues(topic).Inc()
+}
+
+func (g *gossipTracer) DeliverMessage(msg *pubsub.Message) {
+	g.psMsgDeliver.WithLabelValues(*msg.Topic).Inc()
+}
+
+func (g *gossipTracer) ValidateMessage(msg *pubsub.Message) {
+	g.psMsgValidate.WithLabelValues(*msg.Topic).Inc()
+}
+
+func (g *gossipTracer) RejectMessage(msg *pubsub.Message, reason string) {
+	g.psMsgReject.WithLabelValues(*msg.Topic, reason).Inc()
+}
+
+func (g *gossipTracer) DuplicateMessage(msg *pubsub.Message) {
+	g.psMsgDuplicate.WithLabelValues(*msg.Topic).Inc()
+}
+
+func (g *gossipTracer) UndeliverableMessage(msg *pubsub.Message) {
+	g.psMsgUndeliverable.WithLabelValues(*msg.Topic).Inc()
+}
+
+func (g *gossipTracer) ThrottlePeer(p peer.ID) {
+	g.psThrottlePeer.WithLabelValues(agentNameFromPeerID(p, g.host.Peerstore())).Inc()
+}
+
+func (g *gossipTracer) RecvRPC(rpc *pubsub.RPC) {
+	g.setRPCMetrics(g.psRPCSubRecv, g.psRPCPubRecv, g.psRPCRecv, rpc)
+}
+
+func (g *gossipTracer) SendRPC(rpc *pubsub.RPC, p peer.ID) {
+	g.setRPCMetrics(g.psRPCSubSent, g.psRPCPubSent, g.psRPCSent, rpc)
+}
+
+func (g *gossipTracer) DropRPC(rpc *pubsub.RPC, p peer.ID) {
+	g.setRPCMetrics(g.psRPCSubDrop, g.psRPCPubDrop, g.psRPCDrop, rpc)
+}
+
+func (g *gossipTracer) setRPCMetrics(subCtr prometheus.Counter, pubCtr, ctrlCtr *prometheus.CounterVec, rpc *pubsub.RPC) {
+	subCtr.Add(float64(len(rpc.Subscriptions)))
+	if rpc.Control != nil {
+		ctrlCtr.WithLabelValues("ihave").Add(float64(len(rpc.Control.Ihave)))
+		ctrlCtr.WithLabelValues("iwant").Add(float64(len(rpc.Control.Iwant)))
+		ctrlCtr.WithLabelValues("graft").Add(float64(len(rpc.Control.Graft)))
+		ctrlCtr.WithLabelValues("prune").Add(float64(len(rpc.Control.Prune)))
+		ctrlCtr.WithLabelValues("idontwant").Add(float64(len(rpc.Control.Idontwant)))
+	}
+
+	for _, msg := range rpc.Publish {
+		pubCtr.WithLabelValues(*msg.Topic).Inc()
+	}
+}
+
+func agentNameFromPeerID(pid peer.ID, store peerstore.Peerstore) string {
```

Comment:
since we use only `.Get(...)` can we change type from `peerstore.Peerstore` to `PeerMetadata`